chalk-solve/src/clauses/builder.rs

```diff
@@ -77,18 +77,13 @@ impl<'me, I: Interner> ClauseBuilder<'me, I> {
             priority,
         };
 
-        if self.binders.len() == 0 {
-            self.clauses
-                .push(ProgramClauseData::Implies(clause).intern(interner));
-        } else {
-            self.clauses.push(
-                ProgramClauseData::ForAll(Binders::new(
-                    VariableKinds::from(interner, self.binders.clone()),
-                    clause,
-                ))
-                .intern(interner),
-            );
-        }
+        self.clauses.push(
+            ProgramClauseData(Binders::new(
+                VariableKinds::from(interner, self.binders.clone()),
+                clause,
```

Comment:
So, we should be careful here. We indeed may need this to be `clause.shifted_in()` if `self.binders.len() == 0`. This is a little bit awkward. The builder I guess should always be assuming a level of binders.